File: example/00003.go
package main

import (
    "fmt"
    "strings"
)

//实现1 字符串
func lengthOfLongestSubstring(s string) int {
    var not_repeat string
    r := 0
    l := len(s)
    for i :=0;i < l;i ++ {
        v := s[i:i+1]
        i_index := strings.Index(not_repeat,v)
        if i_index > -1 {
            not_repeat = not_repeat[i_index+1:]
            not_repeat += v
        } else {
            not_repeat += v
            n_l := len(not_repeat)
            if n_l > r {
                r = n_l
            }
        }
    }
    return r
}
//实现2
func lengthOfLongestSubstring2(s string) int {
    not_repeat := []byte{}
    r := 0
    l := len(s)
    for i :=0;i < l;i ++ {
        v := s[i]
        //v_i := indexOf(not_repeat,v)
        v_i := -1
        for r_i, r_v := range not_repeat {
            if r_v == v {
                v_i = r_i
                break
            }
        }
        if v_i > -1 {
            not_repeat = not_repeat[v_i+1:]
            not_repeat = append(not_repeat,v)
        } else {
           not_repeat = append(not_repeat,v)
           n_l := len(not_repeat)
           if n_l > r {
               r = n_l
           }
        }

    }
    return r
}

func indexOf(l []byte, s byte) int {
    r := -1
    for i, v := range l {
        if v == s {
            r = i
            break
        }
    }
    return r
}


//实现3
func lengthOfLongestSubstring3(s string) int {
    // 哈希集合，记录每个字符是否出现过
    m := map[byte]int{}
    n := len(s)
    // 右指针，初始值为 -1，相当于我们在字符串的左边界的左侧，还没有开始移动
    rk, ans := -1, 0
    for i := 0; i < n; i++ {
        if i != 0 {
            // 左指针向右移动一格，移除一个字符
            delete(m, s[i-1])
        }
        for rk + 1 < n && m[s[rk+1]] == 0 {
            // 不断地移动右指针
            m[s[rk+1]]++
            rk++
        }
        // 第 i 到 rk 个字符是一个极长的无重复字符子串
        ans = max(ans, rk - i + 1)
    }
    return ans
}

func max(x, y int) int {
    if x < y {
        return y
    }
    return x
}

func main() {
   //s1 := "abcabcbb"
   var s2 string = "aaaab"
   //fmt.Println(lengthOfLongestSubstring2(s1))
   fmt.Println(lengthOfLongestSubstring2(s2))
}

File: example/00003.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        :type s: str
        :rtype: int
        """
        # 最大字符
        r = 0
        # 不重复字符串的数组
        not_repeat = []
        for i in range(len(s)):
            v = s[i]
            if v in not_repeat:
                #如果存在重复元素，删除到重复的位置，重新开始计算
                del not_repeat[0:not_repeat.index(v) + 1]
                not_repeat.append(v)
            else:
                #不存在重复元素，计算最大长度并且保存长度
                not_repeat.append(v)
                l = len(not_repeat)
                if l > r:
                    r = l
        return r
# 实现2
class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # 哈希集合，记录每个字符是否出现过
        occ = set()
        n = len(s)
        # 右指针，初始值为 -1，相当于我们在字符串的左边界的左侧，还没有开始移动
        rk, ans = -1, 0
        for i in range(n):
            if i != 0:
                # 左指针向右移动一格，移除一个字符
                occ.remove(s[i - 1])
            while rk + 1 < n and s[rk + 1] not in occ:
                # 不断地移动右指针
                occ.add(s[rk + 1])
                rk += 1
            # 第 i 到 rk 个字符是一个极长的无重复字符子串
            ans = max(ans, rk - i + 1)
        return ans


s1 = "abcabcbb"
s2 = 'pwwkew'

print(Solution().lengthOfLongestSubstring(s1))
print(Solution().lengthOfLongestSubstring(s2))
print(Solution2().lengthOfLongestSubstring(s1))
print(Solution2().lengthOfLongestSubstring(s2))

